src/server/conn.rs

```diff
@@ -0,0 +1,180 @@
+use crate::{rt::tokio_executor::TokioExecutor, util::PrependAsyncRead};
+use http::{Request, Response};
+use http_body::Body;
+use hyper::{
+    body::Incoming,
+    server::conn::{http1, http2},
+    service::Service,
+};
+use std::{error::Error as StdError, io::Cursor, marker::Unpin};
+use tokio::io::{AsyncRead, AsyncReadExt, AsyncWrite};
+
+const H2_PREFACE: &[u8] = b"PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n";
+
+/// Http1 or Http2 connection.
+pub struct AutoConnection {
+    h1: http1::Builder,
+    h2: http2::Builder<TokioExecutor>,
+}
+
+impl AutoConnection {
+    /// Create a new AutoConn.
+    pub fn new() -> Self {
+        Self {
+            h1: http1::Builder::new(),
+            h2: http2::Builder::new(TokioExecutor::new()),
+        }
+    }
+
+    /// Http1 configuration.
+    pub fn h1(&mut self) -> &mut http1::Builder {
```

Comment:
Perhaps we should use the longer names, `http1()` and `http2()`?

Hm, that _might_ be a case that the individual types should remove the stutter from their methods, so that `auto.http1().http1_writev(true)` were just `auto.http1().writev(true)`...

Then again, it wouldn't be possible to chain configuration. You'd almost want this kind of API:

```rust
auto
    .http1().writev(true)
    .http1().max_read_buf(32 * 1024)
    .http2().max_reset_streams(12)
    // this should use the `auto` serve, not the http2 serve
    .serve_connection(io, svc)
```

Reply:
I think going for the builder api is the way.